### DISCUSSION TITLE
Advice / Best Practices / general questions

### DISCUSSION THREAD

**Discussion by Underknowledge** [kanidm/kanidm#2839] [Q&A]
So, after successfully setting the server part up, excited to start integrating Kanidm. 
My Kani server on a remote "testing" server. I have a few questions as I move forward. 
I guess most of my questions revolve around that I setup kani public facing. 

Audit Trails:
I want to delegate some access (awesome feature!), but I would like to keep track who made changes. Possible months later. 
I've found the Journal logs and Tracing, but they don't seem to provide audit trails like 
user X gave user Y access to group Z, or 
serviceaccount H (not yet sure how idm_people_on_boarding should look like) created the user N. 
I'm looking for. Is there a better way to audit these activities?

Network Co-location:
Should Kanidm be co-located within the same network as Prometheus and other services, or is it fine/ment to be keept it in the wild wide web?

LDAP Exposure:
Do I need to expose the LDAP interface? I prefer not to have unregistered entities see a list of users. 
From my limited understanding, it seems that the Kanidm daemon might not utilize the LDAP part, but rather the API. 
But also on the API I'm not sure if you can just pull a list of users. From what I seen so far there is no auth needed from the clients. 

Client Allowlist:
rel to LDAP Exposure, Is there a built-in allowlist feature for the clients? 
Sure, its nice that I could use it with a server I might have in nord Korea, bur I prefer to control access more tightly without relying on another tool. 
I know firewall rules can manage this, but even I occasionally deactivated them and might have forgotten to re-enable... shame on me.

RPM packages: 
First test host, AlmaLinux 8.5
Now I try to install the client tools, and of cause they're not in the repository. 
ldd tells me that there might be glibc problems. I have no clue bout rust, from my 2 minute search you can build with musl. But yea, honestly no clue. 
Is this already a "last resort" job for SSSD? I dont think that I can convince sshd to spin up a container to validate a SSH key. 

puhh, Sorry for the braindump :dancers:  
Not yet integrated a single client, that's just setting up the server and lurking thou the book. 
cheers!

**Comment by Firstyear** [comment]
> So, after successfully setting the server part up, excited to start integrating Kanidm. My Kani server on a remote "testing" server. I have a few questions as I move forward. I guess most of my questions revolve around that I setup kani public facing.
> 
> Audit Trails: I want to delegate some access (awesome feature!), but I would like to keep track who made changes. Possible months later. I've found the Journal logs and Tracing, but they don't seem to provide audit trails like user X gave user Y access to group Z, or serviceaccount H (not yet sure how idm_people_on_boarding should look like) created the user N. I'm looking for. Is there a better way to audit these activities?

https://github.com/kanidm/kanidm/issues/22

It's certainly on our today, but currently we haven't setup a good infrastructure for it. 

> 
> Network Co-location: Should Kanidm be co-located within the same network as Prometheus and other services, or is it fine/ment to be keept it in the wild wide web?

That's ultimately up to you :) 

> 
> LDAP Exposure: Do I need to expose the LDAP interface? I prefer not to have unregistered entities see a list of users. From my limited understanding, it seems that the Kanidm daemon might not utilize the LDAP part, but rather the API. But also on the API I'm not sure if you can just pull a list of users. From what I seen so far there is no auth needed from the clients.

Unless you need to use the LDAP interface, no you don't need it. 

> 
> Client Allowlist: rel to LDAP Exposure, Is there a built-in allowlist feature for the clients? Sure, its nice that I could use it with a server I might have in nord Korea, bur I prefer to control access more tightly without relying on another tool. I know firewall rules can manage this, but even I occasionally deactivated them and might have forgotten to re-enable... shame on me.

https://github.com/kanidm/kanidm/issues/611

We have most of the plumbing there, we just need to finish it. It's likely this will be done sooner than later. 

In the mean time, you can also disable the `idm_acp_all_accounts_posix_read` rule which prevents these anonymous access. 

> 
> RPM packages: First test host, AlmaLinux 8.5 Now I try to install the client tools, and of cause they're not in the repository. ldd tells me that there might be glibc problems. I have no clue bout rust, from my 2 minute search you can build with musl. But yea, honestly no clue. Is this already a "last resort" job for SSSD? I dont think that I can convince sshd to spin up a container to validate a SSH key.

Is alma based on centos 8? The problem we've hit before is that centos tends to have lagging rust releases to it's been hard to build for. 

If you end up needing to use SSSD, you'll need to use the LDAP interface. That might solve your "anonymous" access problem though because you can create a service-account token for SSSD instances allowing them to read from LDAP while disabling anonymous posix read. 

> 
> puhh, Sorry for the braindump 👯 Not yet integrated a single client, that's just setting up the server and lurking thou the book. cheers!

**Reply by Underknowledge** [reply]
> > Audit Trails: 
> #22
> It's certainly on our today, but currently we haven't setup a good infrastructure for it.

(Ignore my rableing here :smile_cat:  just thoughts at the top of my head)
> Lurking throu it, not sure if open telemetry is a good for such a usecase..
> The generic way would be to store these informations in the DB, but as far as I understood the sqlite db identifys as etcd.
> afaik thats not meant to be a logging backend. Isnt that even sored on a TSDB? 
> I seldom hold such data longer then 3 months. 
> extracting that data then from opentelemetry `that_just_soundslike_slavery_with_extra_steps.jpg`
> Even with replication across multiple machines, I would be fine with a simple logfile on disk.  
> When the need arises then I have to go thou 2 log files, oh my. 


> > Network Co-location: 
> That's ultimately up to you :)

I hear no objections :smirk: 
Well, having it public is already neat for things that are tucked away in their own little networks.
Its rather the question of, how comfortable I feel to just expose a IDM server. it should be fiiiine :tm: 

> > LDAP Exposure
> Unless you need to use the LDAP interface, no you don't need it.

much saaad, already have to do :( 

> > Client Allowlist:
> #611
> We have most of the plumbing there, we just need to finish it. It's likely this will be done sooner than later.
> In the mean time, you can also disable the `idm_acp_all_accounts_posix_read` rule which prevents these anonymous access.
> If you end up needing to use SSSD, you'll need to use the LDAP interface. That might solve your "anonymous" access problem though because you can create a service-account token for SSSD instances allowing them to read from LDAP while disabling anonymous posix read.

I recall an example of toggling rules from the book, but as an LDAP novice, I found it challenging to follow. I attempted to set up a service token but ended up with setting up a sync token what is in hindsight a dumb idea. 
Including a SSSD setup without anonymous read access in the book could be nice. Even with only best effort setup.
would fit the secure by design philosophy.
(maybe I can do a PR for this?)


> > RPM packages: 
> Is alma based on centos 8? The problem we've hit before is that centos tends to have lagging rust releases to it's been hard to build for.

Ever had a play with static binaries? Whenever I'm not feeling well, I dabble in scripting with Golang  (aka, I'm not a developer - I'm clueless!) I just set `CGO_ENABLED=0` and on the other end a binary falls out that I can use across all x86 systems. 

Without dynamic linked system library dependencies, the client tools could be used on most Linux distributions without compatibility issues. An `.rpm` is basically a fancier tar file with metadata.
I'm not sure about `.deb` files, but from the releases I've seen, it looks like you've already solved this problem.
Or you could also just use a tar file and provide an uninstall script, like Rancher or the Determinate Nix Installer does.

So yea, when possible static binarys for the Client tools would be looooveley. Else us poor folks with older systems have to use LDAP :vomiting_face:

**Reply by yaleman** [reply]
Statically building the client tools still wouldn't get you all the way to avoiding LDAP, given that you need the PAM/NSS modules to be compiled for your system either way.

**Reply by Firstyear** [reply]
> > > Client Allowlist:
> > > #611
> > > We have most of the plumbing there, we just need to finish it. It's likely this will be done sooner than later.
> > > In the mean time, you can also disable the `idm_acp_all_accounts_posix_read` rule which prevents these anonymous access.
> > > If you end up needing to use SSSD, you'll need to use the LDAP interface. That might solve your "anonymous" access problem though because you can create a service-account token for SSSD instances allowing them to read from LDAP while disabling anonymous posix read.
> 
> I recall an example of toggling rules from the book, but as an LDAP novice, I found it challenging to follow. I attempted to set up a service token but ended up with setting up a sync token what is in hindsight a dumb idea. Including a SSSD setup without anonymous read access in the book could be nice. Even with only best effort setup. would fit the secure by design philosophy. (maybe I can do a PR for this?)

It would be good to see, yep :) 

> 
> > > RPM packages:
> > > Is alma based on centos 8? The problem we've hit before is that centos tends to have lagging rust releases to it's been hard to build for.
> 
> Ever had a play with static binaries? Whenever I'm not feeling well, I dabble in scripting with Golang (aka, I'm not a developer - I'm clueless!) I just set `CGO_ENABLED=0` and on the other end a binary falls out that I can use across all x86 systems.

all rust binaries are staticly linked, the exception here being we use openssl and sqlite as dylibs. So that makes it ... harder. We also can't statically vendor these simply because openssl vendorede is 1.1 andwe need 3.0.

Finally, the issue as much centos and it's "lagging" with rust tool chains.

**Comment by pmarr** [comment]
RE: RPM packages
Adding steps here for when other rhel8 or enterprise linux clones (rockylinux, alma, centos) users attempt to build the unix_integration pieces.

The min supported rust version at the time was 1.80 while rhel8 only packages up to 1.79, hence the rustup-init.

While the rust-openssl supports vendoring 3.x.x I could not figure out how to get my build-deps to use the vendored version over the systems 1.1.1k, I could compile webauthn-rs with a vendored version.

## unix integration

### compiling for redhat enterprise linux 8

Deps for webauthn: pcsc udev clang+gcc

        curl -LO "https://static.rust-lang.org/rustup/dist/x86_64-unknown-linux-gnu/rustup-init"
        chmod 700 rustup-init
        ./rustup-init
        sudo dnf install pcsc-lite-devel systemd-devel clang-devel
        git clone https://github.com/kanidm/kanidm.git
        git clone https://github.com/kanidm/webauthn-rs.git
        cd webauthn-rs
        KANIDM_BUILD_PROFILE=release_linux cargo build --release
        cd ../kanidm
        KANIDM_BUILD_PROFILE=release_linux make release/kanidm-unix

Changes in webauthn-rs Cargo.toml - Build vendored openssl 3 from source

```diff
-openssl = "^0.10.56"
+openssl = { version = "^0.10.56", features = ["vendored"] }
```

Changes in kanidm Cargo.toml - Use locally compiled webauthn-rs

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 588241cc1..328ed44ef 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -112,11 +112,11 @@ codegen-units = 256
 # ldap3_proto = { git = "https://github.com/kanidm/ldap3.git" }

 # base64urlsafedata = { path = "../webauthn-rs/base64urlsafedata" }
-# webauthn-authenticator-rs = { path = "../webauthn-rs/webauthn-authenticator-rs" }
-# webauthn-rs = { path = "../webauthn-rs/webauthn-rs" }
-# webauthn-rs-core = { path = "../webauthn-rs/webauthn-rs-core" }
-# webauthn-rs-proto = { path = "../webauthn-rs/webauthn-rs-proto" }
-# sshkey-attest = { path = "../webauthn-rs/sshkey-attest" }
+ webauthn-authenticator-rs = { path = "../webauthn-rs/webauthn-authenticator-rs" }
+ webauthn-rs = { path = "../webauthn-rs/webauthn-rs" }
+ webauthn-rs-core = { path = "../webauthn-rs/webauthn-rs-core" }
+ webauthn-rs-proto = { path = "../webauthn-rs/webauthn-rs-proto" }
+ sshkey-attest = { path = "../webauthn-rs/sshkey-attest" }

 # kanidm-hsm-crypto = { path = "../hsm-crypto" }
```
## testing the binaries

Compilation was against the master branch @ commit 537d6fd93b. I'll test again when a 1.6 release cuts.

```
$ ./kanidm_unixd_tasks --version
kanidm_unixd_tasks 1.6.0-dev
$ ./kanidm_unixd --version
kanidm_unixd 1.6.0-dev
$ ./kanidm-unix version
kanidm-unix 1.6.0-dev 537d6fd93b

**Reply by Firstyear** [reply]
@pmarr can you raise an issue on kanidm and webauthn-rs that if we set the openssl min higher, we can vendor (since openssl vendoring is now finally 3.x)
